packages/actions/test/e2e/01-setup.test.ts

```diff
@@ -1,58 +1,225 @@
-    it("should return true for an existing object inside a bucket", async () => {})
-    it("should return false for an non existing object inside a bucket", async () => {})
-    it("should correctly estimate PoT given the number of constraints", async () => {})
-
-    // afterAll(async () => {
-    //     // Clean ceremony and user from DB.
-    //     await adminFirestore.collection("users").doc(userId).delete()
-
-    //     // Remove Auth user.
-    //     await adminAuth.deleteUser(userId)
-
-    //     // Delete admin app.
-    //     await deleteAdminApp()
-    // })
+    // Sample data for running the test.
+    const user = fakeUsersData.fakeUser2
+    const coordinatorEmail = "coordinator@coordinator.com"
+    // storing the uid so we can delete the user after the test
+    let coordinatorUid: string
+
+    // generate passwords for user and coordinator
+    const userPwd = generatePseudoRandomStringOfNumbers(24)
+    const coordinatorPwd = generatePseudoRandomStringOfNumbers(24)
+
+    // Initialize user and admin services.
+    const { userApp, userFunctions } = initializeUserServices()
+    const { adminFirestore, adminAuth } = initializeAdminServices()
+    const userAuth = getAuth(userApp)
+
+    // Get configs for storage.
+    const { ceremonyBucketPostfix } = getStorageConfiguration()
+    const ceremonyData = fakeCeremoniesData.fakeCeremonyScheduledFixed
+    const bucketName = getBucketName(ceremonyData.data.prefix, ceremonyBucketPostfix)
+    // file to upload
+    const localPath = "/tmp/circuitMetadata.log"
+    fs.writeFileSync(localPath, "test data")
+    const objectName = "zkey.zkey"
+    const duplicateBucketName = randomBytes(10).toString("hex")
+
+    beforeAll(async () => {
+        // create a new user without contributor privileges
+        await createNewFirebaseUserWithEmailAndPw(userApp, user.data.email, userPwd)
+        await sleep(5000)
+
+        // Retrieve the current auth user in Firebase.
+        const currentAuthenticatedUser = getCurrentFirebaseAuthUser(userApp)
+        user.uid = currentAuthenticatedUser.uid
+
+        // create account for coordinator
+        await createNewFirebaseUserWithEmailAndPw(userApp, coordinatorEmail, coordinatorPwd)
+        await sleep(5000)
+
+        const currentAuthenticatedCoordinator = getCurrentFirebaseAuthUser(userApp)
+        coordinatorUid = currentAuthenticatedCoordinator.uid
+
+        // add custom claims for coordinator
+        await setCustomClaims(adminAuth, coordinatorUid, { coordinator: true })
+    })
+
+    it("should fail to create a sample ceremony without being a coordinator", async () => {
+        await signInWithEmailAndPassword(userAuth, user.data.email, userPwd)
+        assert.isRejected(createS3Bucket(userFunctions, bucketName))
+    })
+
+    // run these tests only in production mode
+    if (envType === TestingEnvironment.PRODUCTION) {
+        it("should revert when trying to create a ceremony with an existing prefix", async () => {
```

Comment:
There is no control other than in the CLI (when ceremony name is requested interactively when running `setup` command) to prevent the duplication of a ceremony. Therefore, we could assume that this test is checking if creating a bucket twice will results in an error. This is one of the mandatory steps to setup a ceremony. On the other hand, there are the Cloud Functions (createCeremony and intEmptyWaitingQueue). 

This would avoid situations where a ceremony could be created on the CF side even without a reserved bucket for artifacts (these two operations are currently nonatomic). Also, this might give us pause and make us consider moving the createS3Bucket functionality directly inside the CF to make the DB write operation + bucket creation atomic. What are your thoughts on this? 

ping to @gurrpi

Reply:
Tough one here. Where exactly would you move the create S3 bucket function? in the multi part upload? 
Maybe the setup ceremony cloud function should check that a bucket was created and that the required data is inside.  At that point, it would fail if not. 

A frontend check we could do is to check whether the bucket name exists already when getting the ceremony name from the user in the setup phase. This way we can inform them and stop it there (although it will stop later when calling the createS3Bucket function). This can of course be bypassed and would only be true for our "frontend/cli" implementation.

Reply:
Looks like the problem here is "How to ensure the uniqueness of the ceremony?". I think rely on S3 bucket name would be a simple approach, since the uniqueness of S3 bucket name is secured by AWS. Then p0tion's responsibility will become only maintain the link securely between ceremony and S3 bucket.

rationale: a ceremony could be considered as a wrapper of S3 bucket (this view will sacrifice agnosticism, but it seems fair for now)

Reply:
One thing to note is, the process should be stopped once a bucket with the same name is found, as uploading files to the same folder might overwrite if they have the same name.

Reply:
> Maybe the setup ceremony cloud function should check that a bucket was created

Also big plus to this. 
p0tion is kinda simple distributed system (communication between GCP and AWS), so the possibility of request failure, this case is to request to S3, is always present.

If it fails after a few retries over a reasonable time interval, it should probably abort from running the setup command and do a cleanup gracefully

Reply:
> Also big plus to this. p0tion is kinda simple distributed system (communication between GCP and AWS), so the possibility of request failure, this case is to request to S3, is always present.

Bravo, you understood what I meant. The CLI (frontend) side controls are there, the backend/CF side controls are missing. I really wanted to emphasize this by saying that that test case was simply half-checking.
 
> If it fails after a few retries over a reasonable time interval, it should probably abort from running the setup command and do a cleanup gracefully

I think testing is also needed for these discussions. BTW, this comment is a great solution.